### PR TITLE
fix(frontend): Vite dev CJS-to-ESM bridge for generated protobufs. Fixes #13016

### LINKED ISSUE #13016: [frontend] npm start broken -- CJS protobuf files incompatible with Vite dev server
## What happened?

Running `npm start` in the frontend shows a blank white screen with:

`Uncaught SyntaxError: The requested module '/src/third_party/mlmd/generated/ml_metadata/proto/metadata_store_pb.js' does not provide an export named 'Artifact'`

## Root cause

The 6 generated protobuf `_pb.js` files use CommonJS (`require()`, `module.exports`, `goog.object.extend(exports, ...)`) but Vite's dev server expects ES Modules. The production build (`npm run build`) is unaffected because Rollup handles CJS via `build.commonjsOptions`.

Affected files:
- `src/third_party/mlmd/generated/ml_metadata/proto/metadata_store_pb.js`
- `src/third_party/mlmd/generated/ml_metadata/proto/metadata_store_service_pb.js`
- `src/third_party/mlmd/generated/ml_metadata/proto/metadata_store_service_grpc_web_pb.js`
- `src/generated/pipeline_spec/pipeline_spec_pb.js`
- `src/generated/pipeline_spec/google/rpc/status_pb.js`
- `src/generated/platform_spec/kubernetes_platform/kubernetes_executor_config_pb.js`

## How to reproduce

```bash
cd frontend
npm start
# Open http://localhost:3000 → blank white screen, SyntaxError in console
```
## Expected behavior
`npm start` should load the UI without errors.

### PR DESCRIPTION
**Pull request by PR3MM**
## Description

`npm start` shows a blank white screen because 6 generated protobuf `_pb.js` files use CommonJS (`require`/`module.exports`/`goog.object.extend`), but Vite's dev server expects ES Modules. The production build (`npm run build`) is unaffected because Rollup handles CJS via `build.commonjsOptions`.

The generated files use Google Closure Library patterns (`goog.exportSymbol`, `goog.object.extend(exports, ...)`) that standard CJS-to-ESM plugins (like `vite-plugin-commonjs`) cannot handle. This PR adds a custom dev-only Vite plugin that rewrites these patterns to ESM during serving.

### How the plugin works

1. **File detection**: Only processes `.js` files inside `third_party/mlmd/generated/` and `src/generated/`. All other files are untouched.
2. **Rewrite `require()`**: Converts `var jspb = require('google-protobuf')` → `import jspb from 'google-protobuf'` and `grpc.web = require('grpc-web')` → `import __cjsImport0 from 'grpc-web'; grpc.web = __cjsImport0;`
3. **Rewrite exports**: Converts `goog.object.extend(exports, proto.ml_metadata)` and `module.exports = proto.ml_metadata` → `export default proto.ml_metadata;` plus individual named exports derived from `goog.exportSymbol` calls (or namespace property assignments for grpc-web files).
4. **Silent failure detection**: If a matched file yields zero CJS pattern matches, the plugin emits a Vite warning so developers know the generated file format may have changed.

The rewritten files only exist in memory during the dev server session. Nothing is written to disk.

### Why not just regenerate the protos as ESM?

That is the permanent fix but requires more work:
- For `pipeline_spec` and `platform_spec`: ESM `.ts` files already exist alongside the CJS files, so it's mostly deleting the CJS files and updating 1 import.
- For MLMD: no ESM alternative exists. Would need adding `ts-proto` to the generation script and migrating ~55 consumer files from the class-based API (`artifact.getId()`) to the plain object API (`artifact.id`). Estimated 2-4 day effort.

This plugin unblocks `npm start` now while the permanent migration is planned.

## Changes

- **`vite-plugins/generated-cjs-bridge.ts`** -- The plugin. Transforms CJS → ESM for 6 generated files. Handles `var`/`let`/`const`, single/double quotes, optional semicolons, `goog.exportSymbol`, `goog.object.extend(exports, ...)`, `module.exports`, and direct namespace property assignments (grpc-web fallback). Only active in dev mode (`apply: 'serve'`).
- **`vite.config.mts`** -- 2 lines added: import the plugin and activate it.
- **`scripts/generated-cjs-bridge.test.ts`** -- 27 unit tests in 3 groups:
  - 4 path filter tests (correct file detection)
  - 12 synthetic pattern tests (each CJS variant produces correct ESM)
  - 12 real-file integration tests (transforms all 6 actual generated files, verifies zero remaining `require()` and valid `export default`)

## Verification

| Check | Result |
|---|---|
| `tsc --noEmit` | 0 errors |
| `npm run build` | Success, identical output hash to master |
| `vitest run` | 1764 pass (27 new + 1737 existing) |
| `npm start` | UI loads, no SyntaxError |

## Blast radius

- Dev-only (`apply: 'serve'`), cannot run during production builds
- Zero new npm dependencies
- Zero modifications to existing source files (only 2 lines added to config)
- Production build output is byte-for-byte identical to master

Fixes #13016

### CODE REVIEW

File: frontend/vite-plugins/generated-cjs-bridge.ts [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+/**
+ * Paths that contain generated CJS protobuf files. If new generated
+ * directories are added, extend this list.
+ */
+const GENERATED_CJS_PATH_MARKERS = ['third_party/mlmd/generated', '/src/generated/'] as const;
+
+export function isGeneratedCjs(id: string): boolean {
+  if (!id.endsWith('.js')) return false;
+  return GENERATED_CJS_PATH_MARKERS.some((marker) => id.includes(marker));
```
[the remark this answers is not in the document]
Comment (PR3MM) [reply frontend/vite-plugins/generated-cjs-bridge.ts]: Applied in the latest push. isGeneratedCjs now strips query strings/hashes and normalizes backslashes before checking. Added tests for both edge cases

**Review by manaswinidas — CHANGES_REQUESTED** [review]
There are some formatting issues - can you resolve that ?

**Comment by PR3MM** [comment]
> There are some formatting issues - can you resolve that ?

Resolved @manaswinidas

**Comment by jeffspahr** [comment]
Nit: `buildExportBlock()` still emits duplicate named exports if a file ever contains both `goog.object.extend(exports, ...)` and `module.exports = ...`. In that case the transform produces duplicate `export const ...` lines, which makes the generated ESM invalid. I don't think this blocks the PR because the 6 generated files currently use only one export style each, but the synthetic dual-export test should probably either dedupe named exports too or assert that the transformed module parses successfully.
